nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedRouterProcessor.java

```diff
@@ -0,0 +1,267 @@
+                        final PushBackRecordSet pushBackSet = new PushBackRecordSet(recordSet);
+                        final Map<Relationship, List<Record>> outgoingRecords = new HashMap<>();
+                        int index = 0;
+
+                        // Reading in records and evaluate script
+                        while (pushBackSet.isAnotherRecord()) {
+                            final Record record = pushBackSet.next();
+                            final Object evaluatedValue = evaluator.evaluate(record, index++);
+
+                            if (evaluatedValue != null && scriptResultType.isInstance(evaluatedValue)) {
+                                final Optional<Relationship> outgoingRelationship = resolveRelationship(scriptResultType.cast(evaluatedValue));
+
+                                if (outgoingRelationship.isPresent()) {
+                                    if (!outgoingRecords.containsKey(outgoingRelationship.get())) {
+                                        outgoingRecords.put(outgoingRelationship.get(), new LinkedList<>());
+                                    }
+
+                                    outgoingRecords.get(outgoingRelationship.get()).add(record);
+                                } else {
+                                    getLogger().debug("Record with evaluated value {} has no outgoing relationship determined", String.valueOf(evaluatedValue));
+                                }
+                            } else {
+                                throw new ProcessException("Script result is not applicable: " + String.valueOf(evaluatedValue));
+                            }
+                        }
+
+                        // Creating and sending outgoing flow files
+                        for (final Map.Entry<Relationship, List<Record>> entry : outgoingRecords.entrySet()) {
+                            final Relationship relationship = entry.getKey();
+                            final List<Record> records = entry.getValue();
+
+                            if (records.isEmpty()) {
+                                getLogger().debug("No outgoing records for relationship {}", relationship.getName());
+                                continue;
+                            }
+
+                            FlowFile outgoingFlowFile = session.create(incomingFlowFile);
+                            final Map<String, String> attributesToAdd = new HashMap<>();
+
+                            try (
+                                final OutputStream out = session.write(outgoingFlowFile);
+                                final RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, out, outgoingFlowFile);
+                            ) {
+                                attributesToAdd.put("mime.type", writer.getMimeType());
+
+                                if (records.size() == 1) {
+                                    final WriteResult writeResult = writer.write(records.get(0));
```

Comment:
WriteResult can be extracted before if, same in ScriptedTransformRecord class row 231.